server/src/main/java/com/thoughtworks/go/config/ConfigReposMaterialParseResultManager.java

```diff
@@ -33,81 +34,152 @@
-            result = getMaterialResult(fingerprint);
+            result = checkForMaterialErrors(fingerprint);
         }
 
-        if (result !=null) {
-            //config repository was parsed, but does not have merge or clone related errors.
-
+        //config repository was parsed, but does not have merge or clone related errors.
+        if (result != null && result.isSuccessful()) {
             HealthStateScope healthStateScope = HealthStateScope.forPartialConfigRepo(fingerprint);
             List<ServerHealthState> serverHealthStates = serverHealthService.filterByScope(healthStateScope);
-            // It should be noted that the server health state and the result can be thought of as the following:
-            // serverHealthState is the current state
-            // result.isSuccessfully() is the state at the last parse
-
-            // here we're checking that the current state is bad, but the previous state was good
-            if (!serverHealthStates.isEmpty() && result.isSuccessful()) {
-                result.setException(represent(serverHealthStates.get(0)));
+            if (!serverHealthStates.isEmpty()) {
+                result.setException(asError(serverHealthStates.get(0)));
 
                 //clear out the good modification, in case good modification is same as of latest parsed modification
                 if (result.getLatestParsedModification().equals(result.getGoodModification())) {
                     result.setGoodModification(null);
                     result.setPartialConfig(null);
                 }
-            // here we're checking that the current state is good, but the previous state was bad
-            } else if (serverHealthStates.isEmpty() && !result.isSuccessful()) {
-                // The issues that caused the previous state to be bad are gone.
-                // An example would be that previously a upstream pipeline didn't exist, but now does
-                // Clearing latest parsed modification to allow for another parse attempt
-                result.setLatestParsedModification(null);
             }
         }
 
         return result;
     }
 
-    private PartialConfigParseResult getMaterialResult(String fingerprint) {
-        HealthStateScope healthStateScope = HealthStateScope.forMaterialConfig(configRepoService.findByFingerprint(fingerprint).getMaterialConfig());
-        List<ServerHealthState> serverHealthStates = serverHealthService.filterByScope(healthStateScope);
-        if (!serverHealthStates.isEmpty()) {
-            return PartialConfigParseResult.parseFailed(null, represent(serverHealthStates.get(0)));
+    /**
+     * Registers the various ConfigChangedListener instances to mark config-repos for reparse on config updates.
+     *
+     * @param configService the {@link GoConfigService} to register events
+     */
+    void attachConfigUpdateListeners(GoConfigService configService) {
```

Comment:
@bdpiparva told me there's a better way to do this. Since inject multiple listener instances, we'll end up with multiple callbacks, once for the entity update and another for the full save.

He'll push a fix, and point you to the right way to do this (for next time you implement a listener for multiple entities :))

Reply:
Commit b30a4b3513f539261c980272ab828b45be0659e1 is to combine all entity change listeners into one.

Reply:
@ketan @bdpiparva awesome, thanks!